diff --git a/src/secnfs/CMakeLists.txt b/src/secnfs/CMakeLists.txt
index 111b88f..c2c863e 100644
--- a/src/secnfs/CMakeLists.txt
+++ b/src/secnfs/CMakeLists.txt
@@ -18,6 +18,13 @@ set(boost_LIBS
 
 add_definitions(-std=c++11)
 
+set(THREAD_SAFETY ON)
+
+if (THREAD_SAFETY)
+  message(STATUS "thread safety annotation check enabled")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wthread-safety")
+endif()
+
 # for include generated protobuf headers
 message(STATUS "binary dir library: " ${CMAKE_CURRENT_BINARY_DIR})
 include_directories(${CMAKE_CURRENT_BINARY_DIR})
